File: src/embeddings.ts
// SPDX-License-Identifier: AGPL-3.0-only
// Copyright (C) 2026 Discover Legal
// This program is free software: you can redistribute it and/or modify it
// under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version. See <https://www.gnu.org/licenses/>.

import OpenAI from "openai";
import { Config } from "./config.js";
import { logger } from "./logger.js";
import type { EmbeddingResult } from "./types.js";

// ─── Providers ────────────────────────────────────────────────────────────────

// OpenAI embeddings (cloud) — used when LOCAL_EMBEDDINGS=false
let _openaiClient: OpenAI | undefined;
function openaiClient(): OpenAI {
  _openaiClient ??= new OpenAI({ apiKey: Config.embeddings.apiKey });
  return _openaiClient;
}

// ─── Ollama local embedding via /api/embed ─────────────────────────────────

interface OllamaEmbedResponse {
  embeddings: number[][];
  model: string;
}

async function ollamaEmbed(texts: string[]): Promise<number[][]> {
  const url = `${Config.local.ollamaUrl}/api/embed`;
  const model = Config.local.localEmbeddingModel;

  const res = await fetch(url, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ model, input: texts }),
  });

  if (!res.ok) {
    const err = await res.text();
    throw new Error(`Ollama embed failed (${res.status}): ${err}`);
  }

  const data = (await res.json()) as OllamaEmbedResponse;
  if (!data.embeddings?.length) {
    throw new Error("Ollama returned empty embeddings");
  }
  return data.embeddings;
}

// ─── Public API ───────────────────────────────────────────────────────────────

export async function embed(text: string): Promise<EmbeddingResult> {
  if (Config.local.localEmbeddings) {
    const [vec] = await ollamaEmbed([text]);
    return { text, embedding: vec, model: Config.local.localEmbeddingModel };
  }

  const response = await openaiClient().embeddings.create({
    model: Config.embeddings.model,
    input: text,
    dimensions: Config.embeddings.dimensions,
  });
  if (!response.data[0]) {
    throw new Error("OpenAI embeddings API returned an empty data array");
  }
  return { text, embedding: response.data[0].embedding, model: Config.embeddings.model };
}

export async function embedBatch(texts: string[]): Promise<EmbeddingResult[]> {
  if (texts.length === 0) return [];

  if (Config.local.localEmbeddings) {
    // Ollama's /api/embed accepts an array input
    const vectors = await ollamaEmbed(texts);
    if (vectors.length !== texts.length) {
      throw new Error(`Ollama embedding count mismatch: got ${vectors.length}, expected ${texts.length}`);
    }
    const model = Config.local.localEmbeddingModel;
    return texts.map((text, i) => ({ text, embedding: vectors[i], model }));
  }

  const response = await openaiClient().embeddings.create({
    model: Config.embeddings.model,
    input: texts,
    dimensions: Config.embeddings.dimensions,
  });
  return response.data.map((item, i) => ({
    text: texts[i],
    embedding: item.embedding,
    model: Config.embeddings.model,
  }));
}

export function cosineSimilarity(a: number[], b: number[]): number {
  if (a.length !== b.length) {
    throw new Error(`cosineSimilarity: vector length mismatch (${a.length} vs ${b.length})`);
  }
  let dot = 0, magA = 0, magB = 0;
  for (let i = 0; i < a.length; i++) {
    dot += a[i] * b[i];
    magA += a[i] * a[i];
    magB += b[i] * b[i];
  }
  const denom = Math.sqrt(magA) * Math.sqrt(magB);
  return denom === 0 ? 0 : dot / denom;
}

File: biglaw-go/internal/embeddings/embeddings.go
// SPDX-License-Identifier: AGPL-3.0-only
// Copyright (C) 2026 Discover Legal

// Embedding generation — OpenAI text-embedding-3-small or Ollama local model.
// Cosine similarity for Need/Offer matching and semantic search.

package embeddings

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"math"
	"net/http"
	"time"

	"github.com/discover-legal/biglaw-go/internal/config"
)

type EmbeddingResult struct {
	Text      string
	Embedding []float32
	Model     string
}

type Client struct {
	cfg    *config.Config
	client *http.Client
}

func NewClient(cfg *config.Config) *Client {
	return &Client{cfg: cfg, client: &http.Client{Timeout: 30 * time.Second}}
}

func (c *Client) Embed(text string) (*EmbeddingResult, error) {
	results, err := c.EmbedBatch([]string{text})
	if err != nil || len(results) == 0 {
		return nil, err
	}
	return &results[0], nil
}

func (c *Client) EmbedBatch(texts []string) ([]EmbeddingResult, error) {
	if c.cfg.Local.LocalEmbeddings {
		return c.embedOllama(texts)
	}
	return c.embedOpenAI(texts)
}

// ─── OpenAI (text-embedding-3-small) ─────────────────────────────────────────

type openAIEmbedRequest struct {
	Input      interface{} `json:"input"`
	Model      string      `json:"model"`
	Dimensions int         `json:"dimensions,omitempty"`
}

type openAIEmbedResponse struct {
	Data []struct {
		Embedding []float32 `json:"embedding"`
		Index     int       `json:"index"`
	} `json:"data"`
}

func (c *Client) embedOpenAI(texts []string) ([]EmbeddingResult, error) {
	req := openAIEmbedRequest{
		Input:      texts,
		Model:      c.cfg.Embeddings.Model,
		Dimensions: c.cfg.Embeddings.Dimensions,
	}
	body, _ := json.Marshal(req)
	httpReq, _ := http.NewRequest("POST", "https://api.openai.com/v1/embeddings", bytes.NewReader(body))
	httpReq.Header.Set("Content-Type", "application/json")
	httpReq.Header.Set("Authorization", "Bearer "+c.cfg.Embeddings.APIKey)

	resp, err := c.client.Do(httpReq)
	if err != nil {
		return nil, fmt.Errorf("openai embed: %w", err)
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		b, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("openai embed HTTP %d: %s", resp.StatusCode, string(b))
	}
	var r openAIEmbedResponse
	if err := json.NewDecoder(resp.Body).Decode(&r); err != nil {
		return nil, fmt.Errorf("openai embed decode: %w", err)
	}
	// Do not assume the API returns embeddings in request order or in full:
	// place each result by its `index` field and validate the count. A short
	// or reordered response would otherwise silently pair texts with the
	// wrong (or zero-valued) vectors and corrupt the comm graph downstream.
	if len(r.Data) != len(texts) {
		return nil, fmt.Errorf("openai embed count mismatch: got %d, expected %d", len(r.Data), len(texts))
	}
	results := make([]EmbeddingResult, len(texts))
	for _, d := range r.Data {
		if d.Index < 0 || d.Index >= len(texts) {
			return nil, fmt.Errorf("openai embed index out of range: %d", d.Index)
		}
		results[d.Index] = EmbeddingResult{
			Text:      texts[d.Index],
			Embedding: d.Embedding,
			Model:     c.cfg.Embeddings.Model,
		}
	}
	return results, nil
}

// ─── Ollama embeddings ────────────────────────────────────────────────────────

type ollamaEmbedRequest struct {
	Model  string `json:"model"`
	Prompt string `json:"prompt"`
}

type ollamaEmbedResponse struct {
	Embedding []float32 `json:"embedding"`
}

func (c *Client) embedOllama(texts []string) ([]EmbeddingResult, error) {
	model := c.cfg.Local.LocalEmbeddingModel
	baseURL := c.cfg.Local.OllamaURL
	results := make([]EmbeddingResult, len(texts))
	for i, text := range texts {
		reqBody, _ := json.Marshal(ollamaEmbedRequest{Model: model, Prompt: text})
		resp, err := c.client.Post(baseURL+"/api/embeddings", "application/json", bytes.NewReader(reqBody))
		if err != nil {
			return nil, fmt.Errorf("ollama embed: %w", err)
		}
		var r ollamaEmbedResponse
		json.NewDecoder(resp.Body).Decode(&r)
		resp.Body.Close()
		results[i] = EmbeddingResult{Text: text, Embedding: r.Embedding, Model: model}
	}
	return results, nil
}

// ─── Cosine similarity ────────────────────────────────────────────────────────

func CosineSimilarity(a, b []float32) float64 {
	if len(a) != len(b) || len(a) == 0 {
		return 0
	}
	var dot, normA, normB float64
	for i := range a {
		dot += float64(a[i]) * float64(b[i])
		normA += float64(a[i]) * float64(a[i])
		normB += float64(b[i]) * float64(b[i])
	}
	if normA == 0 || normB == 0 {
		return 0
	}
	return dot / (math.Sqrt(normA) * math.Sqrt(normB))
}
